### PR TITLE
Completed a very simple TinyLFU implementation based on conservative update sketch

### PR DESCRIPTION
**Pull request by gilga1983**
now TinyLFU is fully implemented and its parameters can be played with.

### CODE REVIEW

File: simulator/src/main/java/com/clearspring/analytics/stream/frequency/CountMinTinyLFU.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package com.clearspring.analytics.stream.frequency;
```
Comment (ben-manes) [inline simulator/src/main/java/com/clearspring/analytics/stream/frequency/CountMinTinyLFU.java]: This doesn't appear to need to be in their library's package. Instead put it in our admission package as package private.

**Comment by ben-manes** [comment]
Please format using [Google Java style](https://github.com/ben-manes/caffeine/wiki/Contribute). You can import it into Eclipse to format your code. Also you should put the copyright header (just copy it from another class).

File: simulator/src/main/java/com/clearspring/analytics/stream/frequency/CountMinTinyLFU.java (hunk 1 above)
Comment (ben-manes) [inline simulator/src/main/java/com/clearspring/analytics/stream/frequency/CountMinTinyLFU.java]: Oh, I see. You do need to do this to access their package private fields.

File: simulator/src/main/java/com/clearspring/analytics/stream/frequency/CountMinTinyLFU.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * A version of the TinyLFU sketch based on a regular conservative update sketch. 
+ * The difference is that anytime the sum of counters reach a predefined values
+ * we divide all counters by 2 in what is called a reset operation. 
+ * 
+ * @author gilga
+ *  [1] TinyLFU: A Highly Efficient Cache Admission Policy
+ *  http://www.cs.technion.ac.il/~gilga/TinyLFU_PDP2014.pdf
+
+ */
+public class CountMinTinyLFU extends ConservativeAddSketch{
```
Comment (ben-manes) [inline simulator/src/main/java/com/clearspring/analytics/stream/frequency/CountMinTinyLFU.java]: `LFU` --> `Lfu` to conform to Java conventions for abbreviations in camel-case names.

File: simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/TinyLfu.java [hunk 3]
```diff
@@ -40,10 +41,10 @@
 
   private final IFrequency sketch;
 
-  public TinyLfu(double epsOfTotalCount, double confidence) {
+  public TinyLfu(double epsOfTotalCount, double confidence, int sampleSize) {
     // These parameters are best described in...
     // https://github.com/twitter/algebird/blob/develop/algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala
-    this.sketch = new ConservativeAddSketch(epsOfTotalCount, confidence, RANDOM_SEED);
+    this.sketch = new CountMinTinyLFU(epsOfTotalCount,confidence,RANDOM_SEED,sampleSize); //new ConservativeAddSketch(epsOfTotalCount, confidence, RANDOM_SEED);
```
Comment (ben-manes) [inline simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/admission/TinyLfu.java]: Please delete the commented out code following (`// new ...`)

**Comment by ben-manes** [comment]
lgtm after you make the minor formatting changes.

The common approach is to either make changes as new commits and then squash them before merging, or amend the existing commit and force push. Most people end up doing the latter because its easier, but a little less safe if you want to revisit your old history.

So you would make the changes, `git add .`, `git commit --amend`, `git push -f`. Then write a comment to ping me that its ready to review. Or if I lgtm it you could merge into master directly, e.g. `git checkout master` `git merge gilga`, `git push`.

Once you have the hang of git, the project's conventions, etc. you don't need to send me pull requests unless you want feedback or want to propose a change outside of the simulator module.

File: simulator/src/main/resources/reference.conf [hunk 4]
```diff
@@ -80,6 +80,7 @@ caffeine.simulator {
     
     eps = "0.0001"
     confidence = "0.99"
+    sampleSize = "5000"
```
Comment (ben-manes) [inline simulator/src/main/resources/reference.conf]: Typesafe Config uses snake case so this should be `sample_size`

**Comment by gilga1983** [comment]
Sure, as long as you are willing to review what I write I'll be very
excited.
I did not have many people to review my code over the years.


On Sun, Aug 16, 2015 at 11:09 PM, Ben Manes <notifications@github.com>
wrote:

> lgtm after you make the minor formatting changes.
>
> The common approach is to either make changes as new commits and then
> squash them before merging, or amend the existing commit and force push.
> Most people end up doing the latter because its easier, but a little less
> safe if you want to revisit your old history.
>
> So you would make the changes, git add ., git commit --amend, git push -f.
> Then write a comment to ping me that its ready to review. Or if I lgtm it
> you could merge into master directly, e.g. git checkout master git merge
> gilga, git push.
>
> Once you have the hang of git, the project's conventions, etc. you don't
> need to send me pull requests unless you want feedback or want to propose a
> change outside of the simulator module.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ben-manes/caffeine/pull/24#issuecomment-131623002>.
>



-- 
Thanks
Gil Einziger (aka Gilga)

**Comment by ben-manes** [comment]
Cleaned up the pull request and squashed the commits.

**Pull request merged**